prometheus_juju_backup_all_exporter/metrics.py

```diff
@@ -0,0 +1,45 @@
+from abc import ABC, abstractmethod
+from logging import getLogger
+
+logger = getLogger(__name__)
+
+
+class MetricBase(ABC):
+    """Abstract base class for creating custom collectors.
+
+    All collector classes should add this class as a mixin class, then subclass
+    from the appropriate metric family that fits your needs. For example, you
+    can have a custom collector that looks like this:
+
+    ```
+    class CustomCollector(MetricBase, GaugeMetricFamily):
+        def __init__(self, *args, **kwargs):
+            super().__init__(*args, **kwargs)
+
+        def _fetch(self, *args, **kwargs):
+            labels = ["test_label", "test_label_2"]
+            value = 10.0
+            return {"labels": labels, "value": value}
+    ```
+    """
+
+    @abstractmethod
+    def _fetch(self):
+        """User defined fetch method used by `self.collect()`."""
+        pass
+
+    def __new__(cls, *args, **kwargs):
```

Comment:
Why we need to use `__new__` instead of `__init__` here?

Reply:
In short, it is used for logging. Subclassing `MetraBase` won't call `__init__` method until you actually create an instance. So the logging appears a bit lag behind. 

If we use `__init__`, you may receive something like
```
Added sample to xxxx
Added sample to xxxx
Created new metric xxxx
```
If we use `__new__`, it won't:
```
Created new metric xxxx
Added sample to xxxx
Added sample to xxxx
```

Reply:
ack